lib/msf/base/sessions/ssh_command_shell_bind.rb

```diff
@@ -187,28 +258,84 @@ def create(params)
       # Notify handlers before we create the socket
       notify_before_socket_create(self, params)
 
-      mutex = Mutex.new
-      condition = ConditionVariable.new
-      ssh_channel = msf_channel = nil
-
       if params.proto == 'tcp'
         if params.server
-          raise ::Rex::BindFailed.new(params.localhost, params.localport, reason: 'TCP server sockets are not supported by SSH sessions.')
-        end
-
-        ssh_channel = @ssh_connection.open_channel('direct-tcpip', :string, params.peerhost, :long, params.peerport, :string, params.localhost, :long, params.localport) do |new_channel|
-          dlog("new direct-tcpip channel opened to #{Rex::Socket.is_ipv6?(params.peerhost) ? '[' + params.peerhost + ']' : params.peerhost}:#{params.peerport}")
-          msf_channel = TcpClientChannel.new(self, @channel_ticker += 1, new_channel, params)
-          mutex.synchronize do
-            condition.signal
-          end
+          sock = create_server_channel(params)
+        else
+          sock = create_client_channel(params)
         end
       elsif params.proto == 'udp'
         raise ::Rex::ConnectionError.new(params.peerhost, params.peerport, reason: 'UDP sockets are not supported by SSH sessions.')
       end
 
-      raise ::Rex::ConnectionError if ssh_channel.nil?
+      raise ::Rex::ConnectionError unless sock
+
+      # Notify now that we've created the socket
+      notify_socket_created(self, sock, params)
 
+      sock
+    end
+
+    def create_server_channel(params)
+      msf_channel = nil
+      completed_event = Rex::Sync::Event.new
+      @ssh_connection.send_global_request('tcpip-forward', :string, params.localhost, :long, params.localport) do |success, response|
+        if success
+          remote_port = params.localport
+          remote_port = response.read_long if remote_port == 0
+          dlog("Remote forwarding from #{params.localhost} established on port #{remote_port}")
+          key = [params.localhost, remote_port]
+          msf_channel = TcpServerChannel.new(params, self, params.localhost, remote_port)
+          @server_channels[key] = msf_channel
```

Comment:
Shouldn't this channel also be tracked in the larger `#channels` attribute? I see here that you want it keyed by host and port so it can be retrieved later, which makes sense. From an perspective external to this session object, if a server channel is created, it seems like it should be tracked with the other channels.

Reply:
I suspect this relates to my above comment; in that Server Channels are different to Client Channels (again, the naming may be working against us?). The Server Channels create Client Channels when a connection is made to the reverse-listening port. I believe the Client Channels created by the Server Channel should be tracked in the #channels attribute, as the existing `TcpClientChannel` constructor does this addition.